trace.go

```diff
@@ -129,6 +140,13 @@ func (t *Trace) Set(key, val interface{}) {
 	t.mtx.Unlock()
 }
 
+// copyFrom replace all key/value on a trace with a new sets of key/value.
+func (t *Trace) copyFrom(s *Trace) {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+	t.vals = s.vals
```

Comment:
So two things:

1. You only grab the mutex on `t`. The mutex on `s` also needs to be grabbed. But be careful, doing something like

```go
t.mtx.Lock()
defer t.mtx.Unlock()
s.mtx.Lock()
defer s.mtx.Unlock()
```

can lead to deadlocks if you somehow call `t.copyFrom(s)` and `s.copyFrom(t)` concurrently.

2. This does not make a copy of the values from `s` still. Modifications to `s` will be present in `t`.

I think this should be fixed by doing:

```go
vals := s.GetAll()
t.mtx.Lock()
defer t.mtx.Unlock()
t.vals = vals
```

Reply:
Oh, I thought you mentioned not to make two copies here. Did you mean not to make copies for both maps from t and s? I misunderstood that part lol
I will fix it now